README.txt
By: cools 
9:00 PM 7/10/2006

This is a custom luaplayer which adds the some of the functions
from Lua Player .17dk_2 and youresam's netlib version, with all
the functions from Lua Player .20 except module loading.

Functions Added:
System Functions:
System.Quit() --Will quit to the Game Menu on 1.5 and to eLoader on 2.XX
System.setCpuSpeedVHigh() --Sets CPU to 333* Mhz
System.setCpuSpeedHigh() --Sets CPU to 300 Mhz
System.setCpuSpeedMed() --Sets CPU to 266 Mhz
System.setCpuSpeedReg() --Sets CPU to 222 Mhz
System.setCpuSpeedLow() --Sets CPU to 200 Mhz
System.setCpuSpeedVLow() --Sets CPU to 100 Mhz

* = It has actually slowed SOME of my games down, so check to see what it does first...

I am hoping to actually let the CPU be set like this: System.setCpuClock(333,333,166) but
that might not happen...

Netlib 2.0 UDP functions:
Socket.udpConnect(host,port)
udpSocket:udpSend(string)
udpSocket:close()
udpSocket:recv()

For the Rest of the Netlib functions please use the netlib.lua file (the
functions netget(id), netclose(), etc. might be included soon...)


All in all this is just something that was a copy and paste project (well with some
minor modifications). Also Serial I/O MIGHT have been disabled.


The only Non-Module Lua Player .20 with System Over/Under Clocking/Quitting Features with Netlib 2.0 Support!


That's All Folks!

release-notes.txt
Cools has just created what he prides as "The only Non-Module Lua Player .20 with System Over/Under Clocking/Quitting Features with Netlib 2.0 Support!". And although this Lua Player .20 mod is more a copy and paste project, you still have to admire cools for some of the minor modifications he's made. This custom lua player adds some of the functions from Lua Player .17dk_2 and youresam's netlib version plus all the functions - except module loading - included in Lua Player .20. There is no Applications Folder in this mod so if you want to have that, just download Lua Player v0.20 (download provided below) and replace the Eboot. Here are the functions added in this custom lua player: System Functions: System.Quit() --Will quit to the Game Menu on 1.5 and to eLoader on 2.XX System.setCpuSpeedVHigh() --Sets CPU to 333 Mhz System.setCpuSpeedHigh() --Sets CPU to 300 Mhz System.setCpuSpeedMed() --Sets CPU to 266 Mhz System.setCpuSpeedReg() --Sets CPU to 222 Mhz System.setCpuSpeedLow() --Sets CPU to 200 Mhz System.setCpuSpeedVLow() --Sets CPU to 100 Mhz Netlib 2.0 UDP functions: Socket.udpConnect(host,port) udpSocket:udpSend(string) udpSocket:close() udpSocket:recv() Cools noted that the function "System.setCpuSpeedVHigh() --Sets CPU to 333 Mhz" has actually slowed some of his games so he asks all who's going to try out this game to let him know if you suffer the same fate. To get of the rest of the Netlib functions, cools advises us to use the netlib.lua file - the developer hopes to add other functions like netget(id) and netclose() pretty soon. Also in the works is setting of CPU to System.setCpuClock(333,333,166), but cools himself doubt it will happen. The good developer warns that this Lua Player v0.20 mod has not been tested on a 1.5 only on 2.00 tiff. Another warning, "Serial I/O MIGHT have been disabled" with this application.


src/sound.h
#ifndef SOUND__H
#define SOUND__H
// Mikmod interface by nevyn <joachimb@gmail.com>
// Based on code by Jim Shaw

// To stop a playing song, it must also be freed. That's why you can't
// preload it and /then/ play it, like with sounds.


#include <psptypes.h>
#include "mikmod.h"


typedef SAMPLE Sound;
typedef int Voice;

// Internals
extern int _mm_errno;
extern BOOL _mm_critical;
extern char *_mm_errmsg[];

extern UWORD md_mode;
extern UBYTE md_musicvolume;    /* volume of song */
extern UBYTE md_sndfxvolume;    /* volume of sound effects */
extern UBYTE md_reverb;         /* 0 = none;  15 = chaos */
extern UBYTE md_pansep;         /* 0 = mono;  128 == 100% (full left/right) */


/**
 * Initialize Mikmod (sound and music).
 */
extern void initMikmod();

/**
 * Deinitialize Mikmod (sound and music)
 */
extern void unloadMikmod();

/**
 * Load a module music file (mod, xm, s3m, it, ...) and play it.
 * Filename must be a full path ("ms0:/psp/...")
 *
 * @pre filename != NULL
 * @param filename - filename of the music file to load
 */
extern void loadAndPlayMusicFile(char* filename, BOOL loop);

/**
 * Stop music.
 */
extern void stopAndUnloadMusic();

/**
 * Load a module wav file. !!NOTE!!: MUST BE MONO!
 * Filename must be a full path ("ms0:/psp/...")
 *
 * @pre filename != NULL
 * @param filename - filename of the wav file to load
 * @return pointer to a new allocated Sound struct, or NULL on failure
 */
extern Sound* loadSound(char* filename);

/**
 * Unload the loaded sound
 *
 * @pre handle != NULL
 * @param handle - the loaded wav file
 */
extern void unloadSound(Sound* handle);


/**
 * Play sample with volume 255, pan 127.
 *
 * @pre handle != NULL
 * @param handle - Handle of the loaded wav file.
 * @return handle of the currently playing voice
 */
extern Voice playSound(Sound* handle);

/**
 * Stop sample.
 *
 * @pre handle != NULL
 * @param handle - Handle of the loaded wav file.
 */
extern void stopSound(Voice handle);

extern void resumeSound(Voice handle, Sound* soundhandle);

void setSoundLooping(Sound *handle, int loopmode, unsigned long loopstart, unsigned long loopend);
unsigned long getSoundLengthInSamples(Sound *handle);
unsigned long getSoundSampleSpeed(Sound *handle);


/**
 * Sets the volume of the specified voice.
 *
 * @pre handle != NULL && vol >= 0 && vol <= 255
 * @param handle - Handle of the currently playing voice.
 */
extern void setVoiceVolume(Voice handle, UWORD vol);

/**
 * Sets the panning of the specified voice. 127 is middle,
 * 0 is full left and 255 is full right.
 *
 * @pre handle != NULL && pan >= 0 && pan <= 255
 * @param handle - Handle of the currently playing voice.
 */
extern void setVoicePanning(Voice handle, ULONG pan);

/**
 * Sets the speed/frequency of the specified voice.
 *
 * @pre handle != NULL && freq >= 0
 * @param handle - Handle of the currently playing voice.
 */
extern void setVoiceFrequency(Voice handle, ULONG freq);


/**
 * Checks whether music is playing at the moment.
 * 
 * @return whether music is playing or not.
 */
extern BOOL musicIsPlaying();


/**
 * Checks whether a specific voice is playing at the moment.
 * 
 * @pre handle != NULL 
 * @param handle - Handle of the currently playing voice.
 * @return whether music is playing or not.
 */
extern BOOL voiceIsPlaying(Voice handle);
 
 
/**
 * Sets the music volume.
 * 
 * @pre (arg > 0 && arg < 128) || arg==9999
 * @param arg is the volume
 * @return new arg
 */
extern unsigned setMusicVolume(unsigned arg);

/**
 * Sets the sample/SFX/wav volume.
 * 
 * @pre (arg > 0 && arg < 128) || arg==9999
 * @param arg is the volume
 * @return new arg
 */
extern unsigned setSFXVolume(unsigned arg);

/**
 * Sets the reverb.
 * 
 * @pre (arg > 0 && arg < 15) || arg==9999
 * @param arg is the amount of reverb.
 * @return new arg
 */
extern unsigned setReverb(unsigned arg);

/**
 * Sets panoramic separation. 0: mono. 128: full separation.
 * 
 * @pre (arg > 0 && arg < 128) || arg==9999
 * @param arg is the separation
 * @return new arg
 */
extern unsigned setPanSep(unsigned arg);

extern void musicPause();
extern void musicResume();

#endif


src/main.cpp
/*
 * Lua Player for PSP
 * ------------------------------------------------------------------------
 * Licensed under the BSD license, see LICENSE for details.
 *
 * Copyright (c) 2005 Frank Buss <fb@frank-buss.de> (aka Shine)
 *
 * Credits:
 *   many thanks to the authors of the PSPSDK from http://forums.ps2dev.org
 *   and to the hints and discussions from #pspdev on freenode.net
 *
 * $Id: main.cpp 290 2005-12-03 08:49:10Z shine $
 */

#include <pspkernel.h>
#include <pspdebug.h>
#include <pspdisplay.h>
#include <pspctrl.h>
#include <pspsdk.h>
#include <psputility.h>
#include <psprtc.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "graphics.h"
#include "sound.h"
#include "luaplayer.h"
#include "sio.h"

/* the boot.lua */
#include "boot.cpp"

// timezone
static char tz[20];
/* Define the module info section */
PSP_MODULE_INFO("LUAPLAYER", 0x1000, 1, 1);
PSP_MAIN_THREAD_ATTR(0);
PSP_MAIN_THREAD_STACK_SIZE_KB(32); /* smaller stack for kernel thread */

// startup path
char path[256];

/* Exit callback */
int exit_callback(int arg1, int arg2, void *common)
{

	// Unload modules
	unloadMikmod();
		
	sceKernelExitGame();
	return 0;
}

/* Callback thread */
int CallbackThread(SceSize args, void *argp)
{
	int cbid;

	cbid = sceKernelCreateCallback("Exit Callback", exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);

	sceKernelSleepThreadCB();

	return 0;
}

/* Sets up the callback thread and returns its thread id */
int SetupCallbacks(void)
{
	int thid = 0;

	thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, 0, 0);
	if(thid >= 0)
	{
		sceKernelStartThread(thid, 0, 0);
	}

	return thid;
}

int nullOutput(const char *buff, int size)
{
	return size;
}

int debugOutput(const char *buff, int size)
{
	static int debugInitialized = 0;
	if(!buff) {
		debugInitialized = 0;
		return 0;
	}
	if (!debugInitialized) {
		disableGraphics();
		pspDebugScreenInit();
		debugInitialized = 1;
	}
	return pspDebugScreenPrintData(buff, size);
}

void initTimezone()
{
	// calculate the difference between UTC and local time
	u64 tick, localTick;
	sceRtcGetCurrentTick(&tick);
	sceRtcConvertUtcToLocalTime(&tick, &localTick);
	int minutesDelta;
	if (tick < localTick) {
		u64 delta = localTick - tick;
		delta /= sceRtcGetTickResolution();
		minutesDelta = delta;
		minutesDelta = -minutesDelta;
	} else {
		u64 delta = tick - localTick;
		delta /= sceRtcGetTickResolution();
		minutesDelta = delta;
	}
	minutesDelta = minutesDelta / 60;

	// calculate the timezone offset
	int tzOffsetAbs = minutesDelta < 0 ? -minutesDelta : minutesDelta;
	int hours = tzOffsetAbs / 60;
	int minutes = tzOffsetAbs - hours * 60;
	sprintf(tz, "GMT%s%02i:%02i", minutesDelta < 0 ? "-" : "+", hours, minutes);
	setenv("TZ", tz, 1);
	tzset();
}

int user_main(SceSize argc, void* argv)
{
	SetupCallbacks();
	initTimezone();

	// init modules
	initGraphics();
	initMikmod();

	// install new output handlers	
	pspDebugInstallStdoutHandler(debugOutput); 
	pspDebugInstallStderrHandler(debugOutput); 

	// execute Lua script (according to boot sequence)
	getcwd(path, 256);
	char* bootStringWith0 = (char*) malloc(size_bootString + 1);
	memcpy(bootStringWith0, bootString, size_bootString);
	bootString[size_bootString] = 0;
	
	const char * errMsg2 = NULL;
	while(1) { // reload on error
		chdir(path); // set base path luaplater/
		clearScreen(0);
		flipScreen();
		clearScreen(0);
        errMsg2 = runScript(bootStringWith0, true);
		if (errMsg2)
		{                                
            debugOutput("error: ", 7);                               
            debugOutput(errMsg2, strlen(errMsg2));  
            debugOutput("\n", 1);   
			debugOutput("Error: No script file found.\n", 29);
		}
		debugOutput("\nPress start to restart\n", 26);

		SceCtrlData pad; int i;
		sceCtrlReadBufferPositive(&pad, 1); 
		for(i = 0; i < 40; i++) sceDisplayWaitVblankStart();
		while(!(pad.Buttons&PSP_CTRL_START)) sceCtrlReadBufferPositive(&pad, 1); 
		
		debugOutput(0,0);
		initGraphics();
	}
	free(bootStringWith0);
	
	// wait until user ends the program
	sceKernelSleepThread();

	return 0;
}

int main(void)
{
	getcwd(path, 256);
	int err = pspSdkLoadInetModules();
	if (err != 0) {
		pspDebugScreenInit();
		pspDebugScreenPrintf("pspSdkLoadInetModules failed with %x\n", err);
	        sceKernelDelayThread(5*1000000); // 5 sec to read error
	}

	// create user thread, tweek stack size here if necessary
	SceUID thid = sceKernelCreateThread("User Mode Thread", user_main,
	    0x11, // default priority
	    256 * 1024, // stack size (256KB is regular default)
	    PSP_THREAD_ATTR_USER, NULL);
	
	// start user thread, then wait for it to do everything else
	sceKernelStartThread(thid, 0, NULL);
	sceKernelWaitThreadEnd(thid, NULL);
	
	sceKernelExitGame();
	return 0;
}

__attribute__((constructor)) void stdoutInit() 
{ 
	pspKernelSetKernelPC();
	pspSdkInstallNoDeviceCheckPatch();
	pspSdkInstallNoPlainModuleCheckPatch();
	pspKernelSetKernelPC();
	pspKernelSetKernelPC();
	pspSdkInstallNoDeviceCheckPatch();
	pspDebugInstallKprintfHandler(NULL);
//	registerSIODriver();

	// ignore startup messages from kernel, but install the tty driver in kernel mode
	pspDebugInstallStdoutHandler(nullOutput); 
} 



src/luaplayer.cpp
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#include <pspkernel.h>
#include <pspdebug.h>
#include <pspdisplay.h>

#include "luaplayer.h"


static lua_State *L;


const char * runScript(const char* script, bool isStringBuffer )
{
	L = lua_open();
	
	// Standard libraries
	luaL_openlibs(L);

	// luasystem.cpp defines our loadlib.
	// luaopen_loadlib(L);
	
	// Modules
	luaSound_init(L);
	luaControls_init(L);
	luaGraphics_init(L);
	lua3D_init(L);
	luaTimer_init(L);
	luaSystem_init(L);
	luaWlan_init(L);
	
	int s = 0;
	 const char * errMsg = NULL;

	if(!isStringBuffer) 
		s = luaL_loadfile(L, script);
	else 
		s = luaL_loadbuffer(L, script, strlen(script), NULL);
		
	if (s == 0) {
		s = lua_pcall(L, 0, LUA_MULTRET, 0);
	}
	if (s) {
		errMsg = lua_tostring(L, -1);
		lua_pop(L, 1); // remove error message
	}
	lua_close(L);
	
	return errMsg;
}


src/luasystem.cpp
#include <stdlib.h>
#include <string.h>
#include <pspkernel.h>

#include <pspusb.h>
#include <pspusbstor.h>
#include <psppower.h>
#include <pspdebug.h>
#include <unistd.h>
#include <sys/stat.h>
#include "luaplayer.h"

#define SIO_IOCTL_SET_BAUD_RATE 1

static int usbActivated = 0;
static SceUID sio_fd = -1;
static const char* sioNotInitialized = "SIO not initialized.";

SceUID irda_fd = -1;

static int lua_getCurrentDirectory(lua_State *L)
{
	char path[256];
	getcwd(path, 256);
	lua_pushstring(L, path);
	
	return 1;
}

static int lua_setCurrentDirectory(lua_State *L)
{
	const char *path = luaL_checkstring(L, 1);
	if(!path) return luaL_error(L, "Argument error: System.currentDirectory(file) takes a filename as string as argument.");

	lua_getCurrentDirectory(L);
	chdir(path);
	
	return 1;
}

static int lua_curdir(lua_State *L) {
	int argc = lua_gettop(L);
	if(argc == 0) return lua_getCurrentDirectory(L);
	if(argc == 1) return lua_setCurrentDirectory(L);
	return luaL_error(L, "Argument error: System.currentDirectory([file]) takes zero or one argument.");
}


// Move g_dir to the stack and MEET CERTAIN DOOM! If the SceIoDirent is found on the STACK instead of among the globals, the PSP WILL CRASH IN A VERY WEIRD WAY. You have been WARNED.
SceIoDirent g_dir;

static int lua_dir(lua_State *L)
{
	int argc = lua_gettop(L);
	if (argc != 0 && argc != 1) return luaL_error(L, "Argument error: System.listDirectory([path]) takes zero or one argument.");
	

	const char *path = "";
	if (argc == 0) {
		path = "";
	} else {
		path = luaL_checkstring(L, 1);
	}
	int fd = sceIoDopen(path);

	if (fd < 0) {
		lua_pushnil(L);  /* return nil */
		return 1;
	}
	lua_newtable(L);
	int i = 1;
	while (sceIoDread(fd, &g_dir) > 0) {
		lua_pushnumber(L, i++);  /* push key for file entry */

		lua_newtable(L);
			lua_pushstring(L, "name");
			lua_pushstring(L, g_dir.d_name);
			lua_settable(L, -3);
	
			lua_pushstring(L, "size");
			lua_pushnumber(L, g_dir.d_stat.st_size);
			lua_settable(L, -3);
	
			lua_pushstring(L, "directory");
			lua_pushboolean(L, FIO_S_ISDIR(g_dir.d_stat.st_mode));
			lua_settable(L, -3);

		lua_settable(L, -3);
	}

	sceIoDclose(fd);

	return 1;  /* table is already on top */
}

static int lua_createDir(lua_State *L)
{
	const char *path = luaL_checkstring(L, 1);
	if(!path) return luaL_error(L, "Argument error: System.createDirectory(directory) takes a directory name as string as argument.");

	mkdir(path, 0777);
	
	return 0;
}

static int lua_removeDir(lua_State *L)
{
	const char *path = luaL_checkstring(L, 1);
	if(!path) return luaL_error(L, "Argument error: System.removeDirectory(directory) takes a directory name as string as argument.");

	rmdir(path);
	
	return 0;
}

static int lua_removeFile(lua_State *L)
{
	const char *path = luaL_checkstring(L, 1);
	if(!path) return luaL_error(L, "Argument error: System.removeFile(filename) takes a filename as string as argument.");

	remove(path);
	
	return 0;
}

static int lua_rename(lua_State *L)
{
	const char *oldName = luaL_checkstring(L, 1);
	const char *newName = luaL_checkstring(L, 2);
	if(!oldName || !newName)
		return luaL_error(L, "Argument error: System.rename(source, destination) takes two filenames as strings as arguments.");

	// TODO: looks like the stdio "rename" doesn't work
	sceIoRename(oldName, newName);
	
	return 0;
}

static int LoadStartModule(char *path)
{
	u32 loadResult;
	u32 startResult;
	int status;

	loadResult = sceKernelLoadModule(path, 0, NULL);
	if (loadResult & 0x80000000)
		return -1;
	else
		startResult =
		sceKernelStartModule(loadResult, 0, NULL, &status, NULL);

	if (loadResult != startResult)
		return -2;

	return 0;
}

static char modulePath[256];

static void setModulePath()
{
	getcwd( modulePath, 256 );
}

extern "C" {
	extern SceUID psploadlib( const char * name, char * init );
	extern void **findFunction( SceUID id, const char * library, const char * name );
	extern int init( lua_State *L);
}

#define LOADMODULE_ARGERROR "Argument error: System.loadModule(module, init) takes a module name and init method as string arguments."

static int lua_loadModule(lua_State *L)
{
	const char *name = luaL_checkstring(L, 1);
	const char *init = luaL_checkstring(L, 2);

	SceUID uid = psploadlib( name, NULL );
	if ( uid >= 0 ) {
		lua_CFunction f = (lua_CFunction) *(findFunction( uid, name, init )); 
		if (f != NULL)
		{
			lua_pushlightuserdata(L,(void*)uid);
			lua_pushcclosure(L,f,1);
			return 1;
		}
	}

	lua_pushnil(L);
	lua_pushstring(L, LOADMODULE_ARGERROR );
	lua_pushstring(L, (uid >= 0) ? "init" : "open");

	return 3;
}

static int lua_usbActivate(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	if (usbActivated) return 0;
	
	static int modulesLoaded = 0;
	if (!modulesLoaded) {
		//start necessary drivers 
		LoadStartModule("flash0:/kd/semawm.prx"); 
		LoadStartModule("flash0:/kd/usbstor.prx"); 
		LoadStartModule("flash0:/kd/usbstormgr.prx"); 
		LoadStartModule("flash0:/kd/usbstorms.prx"); 
		LoadStartModule("flash0:/kd/usbstorboot.prx"); 
	
		//setup USB drivers 
		int retVal = sceUsbStart(PSP_USBBUS_DRIVERNAME, 0, 0); 
		if (retVal != 0) { 
			printf("Error starting USB Bus driver (0x%08X)\n", retVal); 
			sceKernelSleepThread(); 
		}
		retVal = sceUsbStart(PSP_USBSTOR_DRIVERNAME, 0, 0); 
		if (retVal != 0) { 
			printf("Error starting USB Mass Storage driver (0x%08X)\n", retVal); 
			sceKernelSleepThread(); 
		} 
		retVal = sceUsbstorBootSetCapacity(0x800000); 
		if (retVal != 0) { 
			printf("Error setting capacity with USB Mass Storage driver (0x%08X)\n", retVal); 
			sceKernelSleepThread(); 
		} 
		retVal = 0; 
		modulesLoaded = 1;
	}
	sceUsbActivate(0x1c8);
	usbActivated = 1;
	return 0;
}

static int lua_usbDeactivate(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	if (!usbActivated) return 0;

	sceUsbDeactivate( 0 );  // what value here?
	usbActivated = 0;
	return 0;
}

// battery functions
 
static int lua_powerIsPowerOnline(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushboolean(L, scePowerIsPowerOnline());
	return 1;
}

static int lua_powerIsBatteryExist(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushboolean(L, scePowerIsBatteryExist());
	return 1;
}

static int lua_powerIsBatteryCharging(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushboolean(L, scePowerIsBatteryCharging());
	return 1;
}

static int lua_powerGetBatteryChargingStatus(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushnumber(L, scePowerGetBatteryChargingStatus());
	return 1;
}

static int lua_powerIsLowBattery(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushboolean(L, scePowerIsLowBattery());
	return 1;
}

static int lua_powerGetBatteryLifePercent(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushnumber(L, scePowerGetBatteryLifePercent());
	return 1;
}

static int lua_powerGetBatteryLifeTime(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushnumber(L, scePowerGetBatteryLifeTime());
	return 1;
}

static int lua_powerGetBatteryTemp(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushnumber(L, scePowerGetBatteryTemp());
	return 1;
}

static int lua_powerGetBatteryVolt(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	lua_pushnumber(L, scePowerGetBatteryVolt());
	return 1;
}

static int lua_powerTick(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "wrong number of arguments");
	scePowerTick(0);
	return 0;
}

static int lua_md5sum(lua_State *L)
{
	size_t size;
	const char *string = luaL_checklstring(L, 1, &size);
	if (!string) return luaL_error(L, "Argument error: System.md5sum(string) takes a string as argument.");

	u8 digest[16];
	sceKernelUtilsMd5Digest((u8*)string, size, digest);
	int i;
	char result[33];
	for (i = 0; i < 16; i++) sprintf(result + 2 * i, "%02x", digest[i]);
	lua_pushstring(L, result);
	
	return 1;
}

static int lua_sioInit(lua_State *L)
{
	if (lua_gettop(L) != 1) return luaL_error(L, "baud rate expected.");
	int baudRate = luaL_checkint(L, 1);
	if (sio_fd < 0) sio_fd = sceIoOpen("sio:", PSP_O_RDWR, 0);
	if (sio_fd < 0) return luaL_error(L, "failed create SIO handle.");
	sceIoIoctl(sio_fd, SIO_IOCTL_SET_BAUD_RATE, &baudRate, sizeof(baudRate), NULL, 0);
	
	return 0;
}

static int lua_sioWrite(lua_State *L)
{
	if (sio_fd < 0) return luaL_error(L, sioNotInitialized);
	size_t size;
	const char *string = luaL_checklstring(L, 1, &size);
	if (!string) return luaL_error(L, "Argument error: System.sioWrite(string) takes a string as argument.");
	sceIoWrite(sio_fd, string, size);
	
	return 0;
}

static int lua_sioRead(lua_State *L)
{
	if (sio_fd < 0) return luaL_error(L, sioNotInitialized);
	if (lua_gettop(L) != 0) return luaL_error(L, "no arguments expected.");
	char data[256];
	int count = sceIoRead(sio_fd, data, 256);
	if (count > 0) {
		lua_pushlstring(L, data, count);
	} else {
		lua_pushstring(L, "");
	}
	
	return 1;
}

static int lua_irdaInit(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "no arguments expected.");
	if (irda_fd < 0) irda_fd = sceIoOpen("irda0:", PSP_O_RDWR, 0);
	if (irda_fd < 0) return luaL_error(L, "failed create IRDA handle.");
	
	return 0;
}

static int lua_irdaWrite(lua_State *L)
{
	if (irda_fd < 0) return luaL_error(L, "irda not initialized");
	size_t size;
	const char *string = luaL_checklstring(L, 1, &size);
	if (!string) return luaL_error(L, "Argument error: System.sioWrite(string) takes a string as argument.");
	sceIoWrite(irda_fd, string, size);
	
	return 0;
}

static int lua_irdaRead(lua_State *L)
{
	if (irda_fd < 0) return luaL_error(L, "irda not initialized");
	if (lua_gettop(L) != 0) return luaL_error(L, "no arguments expected.");
	char data[256];
	int count = sceIoRead(irda_fd, &data, 256);
	if (count > 0) {
		lua_pushlstring(L, data, count);
	} else {
		lua_pushstring(L, "");
	}
	
	return 1;
}

static int lua_sleep(lua_State *L)
{
	if (lua_gettop(L) != 1) return luaL_error(L, "milliseconds expected.");
	int milliseconds = luaL_checkint(L, 1);
	sceKernelDelayThread(milliseconds * 1000);
	return 0;
}

static int lua_getFreeMemory(lua_State *L)
{
	if (lua_gettop(L) != 0) return luaL_error(L, "no arguments expected.");
	
	// for a realistic number, try to allocate 1 mb blocks until out of memory
	void* buf[64];
	int i = 0;
	for (i = 0; i < 64; i++) {
		buf[i] = malloc(1024 * 1024);
		if (!buf[i]) break;
	}
	int result = i;
	for (; i >= 0; i--) {
		free(buf[i]);
	}
	lua_pushnumber(L, result * 1024 * 1024);
	return 1;
}

static int lua_Quit(lua_State *L)
{
	sceKernelExitGame();
	return 0;
}

static int lua_setCpuSpeedVHigh(lua_State *L)
{
	scePowerSetClockFrequency(333, 333, 166);
	return 0;
}

static int lua_setCpuSpeedHigh(lua_State *L)
{
	scePowerSetClockFrequency(300, 300, 150);
	return 0;
}

static int lua_setCpuSpeedMed(lua_State *L)
{
	scePowerSetClockFrequency(266, 266, 133);
	return 0;
}

static int lua_setCpuSpeedReg(lua_State *L)
{
	scePowerSetClockFrequency(222, 222, 111);
	return 0;
}

static int lua_setCpuSpeedLow(lua_State *L)
{
	scePowerSetClockFrequency(200, 200, 100);
	return 0;
}

static int lua_setCpuSpeedVLow(lua_State *L)
{
	scePowerSetClockFrequency(100, 100, 50);
	return 0;
}

static const luaL_reg System_functions[] = {
  {"currentDirectory",              lua_curdir},
  {"listDirectory",           	    lua_dir},
  {"createDirectory",               lua_createDir},
  {"removeDirectory",               lua_removeDir},
  {"removeFile",                    lua_removeFile},
  {"rename",                        lua_rename},
  {"usbDiskModeActivate",           lua_usbActivate},
  {"usbDiskModeDeactivate",    	    lua_usbDeactivate},
  {"powerIsPowerOnline",            lua_powerIsPowerOnline},
  {"powerIsBatteryExist",           lua_powerIsBatteryExist},
  {"powerIsBatteryCharging",        lua_powerIsBatteryCharging},
  {"powerGetBatteryChargingStatus", lua_powerGetBatteryChargingStatus},
  {"powerIsLowBattery",             lua_powerIsLowBattery},
  {"powerGetBatteryLifePercent",    lua_powerGetBatteryLifePercent},
  {"powerGetBatteryLifeTime",       lua_powerGetBatteryLifeTime},
  {"powerGetBatteryTemp",           lua_powerGetBatteryTemp},
  {"powerGetBatteryVolt",           lua_powerGetBatteryVolt},
  {"powerTick",                     lua_powerTick},
  {"md5sum",                        lua_md5sum},
  {"sioInit",                       lua_sioInit},
  {"sioRead",                       lua_sioRead},
  {"sioWrite",                      lua_sioWrite},
  {"irdaInit",                      lua_irdaInit},
  {"irdaRead",                      lua_irdaRead},
  {"irdaWrite",                     lua_irdaWrite},
  {"sleep",                         lua_sleep},
  {"getFreeMemory",                 lua_getFreeMemory},
  {"Quit",                          lua_Quit},
  {"setCpuSpeedVHigh",              lua_setCpuSpeedVHigh},
  {"setCpuSpeedHigh",               lua_setCpuSpeedHigh},
  {"setCpuSpeedMed",                lua_setCpuSpeedMed},
  {"setCpuSpeedReg",                lua_setCpuSpeedReg},
  {"setCpuSpeedLow",                lua_setCpuSpeedLow},
  {"setCpuSpeedVLow",               lua_setCpuSpeedVLow},
  {0, 0}
};
void luaSystem_init(lua_State *L) {
	setModulePath();
	luaL_openlib(L, "System", System_functions, 0);
	lua_register(L, "loadlib", lua_loadModule);
}

